### PR TITLE
Refactor RBAC for project and cluster level catalog

### PR DESCRIPTION
**Pull request by mrajashree**
https://github.com/rancher/rancher/issues/16316

Types PR: https://github.com/rancher/types/pull/641

### CODE REVIEW

**Comment by alena1108** [comment]
@mrajashree could you remove that extra commit that you've added to make the build work?

File: app/role_data.go [hunk 1]
```diff
@@ -64,17 +65,17 @@ func addRoles(management *config.ManagementContext) (string, error) {
 	rb = newRoleBuilder()
 
 	// K8s default roles
-	rb.addRoleTemplate("Kubernetes cluster-admin", "cluster-admin", "cluster", true, true, true)
-	rb.addRoleTemplate("Kubernetes admin", "admin", "project", true, true, true)
-	rb.addRoleTemplate("Kubernetes edit", "edit", "project", true, true, true)
-	rb.addRoleTemplate("Kubernetes view", "view", "project", true, true, true)
+	rb.addRoleTemplate("Kubernetes cluster-admin", "cluster-admin", "cluster", true, true, true, true)
```
Comment (alena1108) [inline app/role_data.go]: @mrajashree do we have a UI ticket filed to support this extra field?

File: pkg/catalog/manager/traverse.go [hunk 2]
```diff
@@ -28,7 +29,14 @@ func (m *Manager) traverseAndUpdate(repoPath, commit string, cmt *CatalogInfo) e
 	}
 
 	// list all existing templates
-	templateMap, err := m.getTemplateMap(catalog.Name)
+	if catalogType == client.CatalogType {
```
Comment (alena1108) [inline pkg/catalog/manager/traverse.go]: Nit: can we consider using case switch instead of if/else

File: pkg/catalog/manager/traverse.go [hunk 3]
```diff
@@ -77,12 +85,21 @@ func (m *Manager) traverseAndUpdate(repoPath, commit string, cmt *CatalogInfo) e
...
 		}
 
-		template := v3.Template{
+		template := v3.CatalogTemplate{
 			ObjectMeta: metav1.ObjectMeta{
 				Name: chart,
 			},
 		}
 
+		if catalogType == client.CatalogType {
```
Comment (alena1108) [inline pkg/catalog/manager/traverse.go]: same here. As it's string comparison, lets switch to case/switch

File: pkg/controllers/management/auth/globalrolebinding_handler.go [hunk 4]
```diff
@@ -129,5 +150,83 @@ func (grb *globalRoleBindingLifecycle) reconcileGlobalRoleBinding(globalRoleBind
...
+			for _, resource := range rule.Resources {
+				if resource == templateResourceRule {
+					templateRuleExists = true
+					catalogTemplateRule = v1.PolicyRule{
+						APIGroups: rule.APIGroups,
+						Resources: []string{catalogTemplateResourceRule},
+						Verbs:     rule.Verbs,
+					}
+				}
+				if resource == templateVersionResourceRule {
```
Comment (alena1108) [inline pkg/controllers/management/auth/globalrolebinding_handler.go]: else if

File: pkg/controllers/management/auth/globalrolebinding_handler.go [hunk 5]
```diff
@@ -51,6 +70,8 @@ func (grb *globalRoleBindingLifecycle) Remove(obj *v3.GlobalRoleBinding) (runtim
 }
 
 func (grb *globalRoleBindingLifecycle) reconcileGlobalRoleBinding(globalRoleBinding *v3.GlobalRoleBinding) error {
+	var templateRuleExists, templateVersionRuleExists bool
```
Comment (alena1108) [inline pkg/controllers/management/auth/globalrolebinding_handler.go]: I don't think there is a need for bool and objects. We can just use pointer type for catalogTemplateRule and catalogTemplateVersionRule if you need to figure out the presence of either. Then you can use & when passing down the value

File: pkg/controllers/management/auth/globalrolebinding_handler.go [hunk 6]
```diff
@@ -129,5 +150,83 @@ func (grb *globalRoleBindingLifecycle) reconcileGlobalRoleBinding(globalRoleBind
...
+						Verbs:     rule.Verbs,
+					}
+				}
+			}
+		}
+	}
+	// If rules for "templates and "templateversions" exists, create a role for the granting access to
+	// catalogtemplates and catalogtemplateversions in the global namespace
+	if templateRuleExists || templateVersionRuleExists {
+		rules := []v1.PolicyRule{}
```
Comment (alena1108) [inline pkg/controllers/management/auth/globalrolebinding_handler.go]: use `var rules []v1.PolicyRule{}` declaration

**Comment by alena1108** [comment]
@mrajashree I've pulled this branch to test, and seeing 2 issues:

* on this branch, the catalog templates icons are not being loaded. Here is the ngrok error:

```
GET /v3/templates/cattle-global-data:library-wordpress/icon          404 Not Found
GET /v3/templates/cattle-global-data:library-vault-operator/icon     404 Not Found
GET /v3/templates/cattle-global-data:library-prometheus/icon         404 Not Found
GET /v3/templates/cattle-global-data:library-portworx/icon           404 Not Found
GET /v3/templates/cattle-global-data:library-phpbb/icon              404 Not Found
GET /v3/templates/cattle-global-data:library-redis/icon              404 Not Found
GET /v3/templates/cattle-global-data:library-openebs/icon            404 Not Found
GET /v3/templates/cattle-global-data:library-nfs-provisioner/icon    404 Not Found
GET /v3/templates/cattle-global-data:library-mysql/icon              404 Not Found
GET /v3/templates/cattle-global-data:library-mongodb-replicaset/icon 404 Not Found
```

and this error in the rancher log `2019/01/14 16:48:07 [ERROR] CatalogController library [catalog] failed with : failed to sync templates. Resetting commit. Multiple error occurred: [Error in HTTP GET of [file://../logo.png], error: Get file://../logo.png: unsupported protocol scheme "file"]`

* rancher upgrade doesn't work. Try upgrading from rancher master to this branch, and you won't see any templates. The message is "can't find any compatible templates"

File: pkg/api/customization/catalog/template.go [hunk 7]
```diff
@@ -26,15 +26,15 @@ func TemplateFormatter(apiContext *types.APIContext, resource *types.RawResource
...
+	if val[client.CatalogTemplateFieldCatalogID] != nil {
 		//catalog link
 		catalogSchema := apiContext.Schemas.Schema(&managementschema.Version, client.CatalogType)
 		catalogName := strings.Split(resource.ID, "-")[0]
 		resource.Links["catalog"] = apiContext.URLBuilder.ResourceLinkByID(catalogSchema, catalogName)
 	}
 
-	if val[client.TemplateFieldProjectCatalogID] != nil {
-		prjCatID, ok := val[client.TemplateFieldProjectCatalogID].(string)
+	if val[client.CatalogTemplateFieldProjectCatalogID] != nil {
```
Comment (alena1108) [inline pkg/api/customization/catalog/template.go]: else if?

File: pkg/api/customization/catalog/template.go [hunk 8]
```diff
@@ -44,7 +44,7 @@ func TemplateFormatter(apiContext *types.APIContext, resource *types.RawResource
 	}
 
 	if val[client.TemplateFieldClusterCatalogID] != nil {
```
Comment (alena1108) [inline pkg/api/customization/catalog/template.go]: else if

**Comment by mrajashree** [comment]
Upgrade issue is resolved. Now templates are present post upgrade, since catalogtemplates are getting created. Same for templateversions. ByID issue is resolved too, and getting the icon and appreadme works too. Role templates didn't require any upgrade logic for the administrative bit, it got set to true correctly on upgrade

**Comment by alena1108** [comment]
LGTM

**Pull request merged**